scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala

```diff
@@ -17,20 +17,109 @@
+import scala.util.Try
 import scala.util.chaining._
 
 object CoderMaterializer {
   import com.spotify.scio.ScioContext
 
-  private[scio] case class CoderOptions(nullableCoders: Boolean, kryo: KryoOptions)
+  private[scio] case class CoderOptions(
+    nullableCoders: Boolean,
+    kryo: KryoOptions,
+    zstdDictMapping: Map[String, Array[Byte]]
+  )
   private[scio] object CoderOptions {
+    private val cache: ConcurrentHashMap[String, CoderOptions] = new ConcurrentHashMap()
+    private val ZstdArgRegex = "([^:]+):(.*)".r
+    private val ZstdPackageBlacklist =
+      List("scala.", "java.", "com.spotify.scio.", "org.apache.beam.")
+
     final def apply(o: PipelineOptions): CoderOptions = {
-      val nullableCoder = o.as(classOf[com.spotify.scio.options.ScioOptions]).getNullableCoders
-      new CoderOptions(nullableCoder, KryoOptions(o))
+      val isTest = Try(o.as(classOf[ApplicationNameOptions])).toOption
+        .exists(o => TestUtil.isTestId(o.getAppName))
+
+      cache.computeIfAbsent(
+        ObjectUtils.identityToString(o),
+        { _ =>
+          val scioOpts = o.as(classOf[com.spotify.scio.options.ScioOptions])
+          scioOpts.getAppArguments
+          val nullableCoder = scioOpts.getNullableCoders
+          val zstdDictPaths = Option(scioOpts.getZstdDictionary)
+            .map(_.asScala)
+            .getOrElse(List.empty)
+            .map {
+              case s @ ZstdArgRegex(className, path) =>
+                if (!isTest) {
+                  ZstdPackageBlacklist.foreach { packagePrefix =>
+                    if (className.startsWith(packagePrefix)) {
+                      throw new IllegalArgumentException(
+                        s"zstdDictionary command-line arguments may not be used for class $className. " +
+                          s"Provide Zstd coders manually instead."
+                      )
+                    }
+                  }
+                }
+
+                try {
+                  // ensure class exists
+                  Class.forName(className)
+                } catch {
+                  case e: ClassNotFoundException =>
+                    throw new IllegalArgumentException(
+                      s"Class for zstdDictionary argument ${s} not found.",
+                      e
+                    )
+                }
+                className.replaceAll("\\$", ".") -> path
+              case s =>
+                throw new IllegalArgumentException(
+                  "zstdDictionary arguments must be in a colon-separated format. " +
+                    s"Example: `com.spotify.ClassName:gs://path`. Found: ${s}"
+                )
+            }
+            .groupBy(_._1)
+            .map { case (className, values) => className -> values.map(_._2).toSet }
+
+          val dupes = zstdDictPaths
+            .collect {
+              case (className, values) if values.size > 1 =>
+                s"Class $className -> [${values.mkString(", ")}]"
+            }
```

Comment:
maybe a `partitionMap` can be nicer so we have a sanitized return type and list of errors.